Implementing rule padded-blocks (no fixer)

Xapphire13 opened the pull request:
Closes #213

Xapphire13 commented:
Ping. Are there any active reviewers around? I noticed the other PRs appear to have gone stale ~1 month back.

buzinas commented:
Sorry @Xapphire13, very busy lately. I'll check this as soon as I can.

@jmlopez-rod if you have any free time, can you check this one, please?

Xapphire13 commented:
No worries at all =]

neilsoult commented:
I would love to see this rule included.

Xapphire13 commented:
@buzinas, @jmlopez-rod

Ping once more =]

MeltingMosaic commented:
@buzinas @jmlopez-rod 
Hey guys, my team is waiting on getting this rule in - are you folks still around?

buzinas commented:
Yeah, I am, but very very busy lately.

Really sorry, guys! I'll try to add some more collaborators to the project.

buzinas commented on src/rules/noInvalidRegexpRule.ts:
```diff
@@ -26,9 +26,8 @@ class NoInvalidRegexpWalker extends Lint.RuleWalker {
-        let regex: RegExp;
         try {
-          regex = new RegExp(expr, flags);
+          (() => new RegExp(expr, flags))();
```
I think we can simply instantiate the Regex here, no need for a IIFE.

Xapphire13 replied:
I agree, not really sure why I had it like that haha.

Xapphire13 replied:
Oh wait, I think the IIFE was to get past linter rules to do with "no unused expressions". Let my change and try it out

Xapphire13 replied:
I can change and suppress if needed

Xapphire13 commented:
@buzinas, even though im a collaborator, I wont have access to publish new versions on NPM will I? Are you still able to handle that?

jmlopez-rod commented on src/rules/paddedBlocksRule.ts:
```diff
@@ -0,0 +1,270 @@
+
+import { forEachComment } from 'tsutils';
+
+const RULE_NAME = 'padded-blocks';
```
Lets change the name to `ter-padded-blocks`. It has happened that tslint takes the names of the rule and they are slightly different from our implementation. In order to avoid surprises we can just leave the two rules living side by side, and later, if we realize that they do the same we can deprecate it if the rule from tslint does the same.

jmlopez-rod commented on src/rules/noInvalidRegexpRule.ts:
```diff
@@ -26,9 +26,9 @@ class NoInvalidRegexpWalker extends Lint.RuleWalker {
-        let regex: RegExp;
         try {
-          regex = new RegExp(expr, flags);
+          // tslint:disable-next-line no-unused-expression
```
My last PR is in conflict. Please leave your change.

Xapphire13 commented:
@jmlopez-rod 

I've addressed your feedback and am giving time for you to give a second look. Will merge tomorrow.

Thank you both so much @buzinas and @jmlopez-rod

jmlopez-rod commented on src/test/rules/terPaddedBlocksRuleTests.ts:
```diff
@@ -0,0 +1,524 @@
+// This package also highlights the entire block when both the start and end are invalid. ESLint
+// creates two separate errors in that case.
+//
+// ESLint Tests: https://github.com/eslint/eslint/blob/master/tests/lib/rules/padded-blocks.js
```
Glad you pointed out the tests from eslint

jmlopez-rod commented:
I've merged it. I'll write a section on the readme with instructions on how to release to npm. It all boils down to doing

```
bash ./bash_scripts/release.sh
```

This will  run the tests and stop you from releasing if they fail. It will also ask you to check that you bumped the version number and that you have written in the changelog. Once I do that I'd like for you to try to make a minor version release (provided that you now have permission on npm).

neilsoult commented:
is this included in the npm 5.0.0 version? It doesn't appear to be included in my node_modules source.

jmlopez-rod commented:
@neilsoult , not yet, I just checked and it seems that only @buzinas and I can do uploads to npm. I'll do the 5.1.0 release this afternoon which will include this rule.
